[dtensor][8/N] switch DeviceMesh to use numpy array for devices

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92126
* __->__ #92931
* #90735
* #90734
* #90733

Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

Differential Revision: [D42876247](https://our.internmc.facebook.com/intern/diff/D42876247)

wanchaol committed:
[dtensor][7/N] switch DeviceMesh to use numpy array for devices

Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

[ghstack-poisoned]

aazzolini commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -38,19 +40,6 @@ def set_global_device_mesh(mesh: Optional["DeviceMesh"]) -> None:
     _global_device_mesh = mesh
 
 
-# We want a type for "can be passed to torch.as_tensor()";
-# this is a recursive sequence type, which isn't fully supported
-# yet in python. This construct simulates that up to depth 7.
-T = TypeVar("T")
-_L = Union[T, Sequence[T]]
-NDIntList = _L[_L[_L[_L[_L[_L[_L[int]]]]]]]
-
-MeshExprT = Union[
-    torch.Tensor,
-    NDIntList,
-]
-
-
```
Thanks this was a bit weird :)

aazzolini reviewed (approved):
Thanks! One suggestion for unit testing would be to create a DeviceMesh in FakeMode to reproduce the issue that I had! Or maybe create a DeviceMesh inside of a simple function and make_fx on it?

wanchaol committed:
Update on "[dtensor][7/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

[ghstack-poisoned]

wanchaol commented:
> Thanks! One suggestion for unit testing would be to create a DeviceMesh in FakeMode to reproduce the issue that I had! Or maybe create a DeviceMesh inside of a simple function and make_fx on it?

Thanks @aazzolini! Just added a test with the FakeTensorMode, it works as expected. However if we want to fully avoid this issue, we might want to stop supporting DeviceMesh taking tensor as input for the mesh field, and use either numpy array or n-d list instead. Let me know if you want to go with that direction :)

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

[ghstack-poisoned]

XilunWu commented:
> > Thanks! One suggestion for unit testing would be to create a DeviceMesh in FakeMode to reproduce the issue that I had! Or maybe create a DeviceMesh inside of a simple function and make_fx on it?
> 
> Thanks @aazzolini! Just added a test with the FakeTensorMode, it works as expected. However if we want to fully avoid this issue, we might want to stop supporting DeviceMesh taking tensor as input for the mesh field, and use either numpy array or n-d list instead. Let me know if you want to go with that direction :)

Would allowing passing Tensor as argument but converting it to numpy array or n-d list in `DeviceMesh._init_` avoid the tracing issue?

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42876247).

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

Differential Revision: [D42876247](https://our.internmc.facebook.com/intern/diff/D42876247)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42876247).

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

Differential Revision: [D42876247](https://our.internmc.facebook.com/intern/diff/D42876247)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

Differential Revision: [D42876247](https://our.internmc.facebook.com/intern/diff/D42876247)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][8/N] switch DeviceMesh to use numpy array for devices"


Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

Differential Revision: [D42876247](https://our.internmc.facebook.com/intern/diff/D42876247)

[ghstack-poisoned]
